bindings/rust/bench/benches/memory.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use criterion::{
+    criterion_group, criterion_main,
+    measurement::{Measurement, ValueFormatter},
+    BenchmarkGroup, Criterion,
+};
+use rand::thread_rng;
+use rand_distr::{Distribution, Normal};
+use std::{fs::read_to_string, path::Path, time::Duration};
+
+fn get_bytes_from_snapshot(name: &str, i: i32) -> i32 {
+    // number of bytes in snapshot starts on 8th line, 12th character
+    read_to_string(format!("target/memory/{name}/{i}.snapshot"))
+        .unwrap()
+        .lines()
+        .nth(7)
+        .unwrap()[11..]
+        .parse()
+        .unwrap()
+}
+
+/// Get the difference in bytes between two snapshots, which is memory of the
+/// `i`th TlsBenchHarness (client and server)
+fn get_bytes_diff(name: &str, i: i32) -> i32 {
+    get_bytes_from_snapshot(name, i + 1) - get_bytes_from_snapshot(name, i)
+}
+
+/// Read in memory snapshots and give Criterion memory values to analyze and plot
+fn read_library(group: &mut BenchmarkGroup<'_, Bytes>, name: &str) {
+    if !Path::new("target/memory/").is_dir() {
+        panic!("need to run memory-bench.sh first");
+    }
+
+    let mut i = 0;
+    let normal = Normal::new(0.0, 10000.0).unwrap();
+    group.bench_function(name, |b| {
+        // iter_custom requires closure returning time taken for `iters` iterations
+        // read in `iters` memory snapshots and return sum of memory taken
+        b.iter_custom(|iters| {
+            let mut sum = 0;
+            for _ in 0..iters {
+                sum += get_bytes_diff(name, i) + normal.sample(&mut thread_rng()) as i32; // add jitter for plotting
+                i = (i + 1) % 100; // read in snapshots again if ran out
```

Comment:
Criterion isn't designed for this, and it runs the function being benched as many times as it thinks it can within around (the default of) 5 seconds. With `iter_custom()`, Criterion expects us to return the time a function takes `iters` times, so to translate to memory benchmarks, we return the memory we expect to use with `iters` connections. To smooth out the distribution (as I mentioned below since there is some variance in `get_bytes_diff()`), the benches just loop through all of the snapshots over and over.

Reply:
I use Criterion to plot memory benchmarks in order for the memory graph to be in the same place as the other graphs, have the same theme, and be in the same report. I guess we could plot it ourselves and figure out how to plot a graph with the same theme and embed it into the HTML report, but the Rust plotting library doesn't seem to be that easy to use; it'd be a lot of time for not that much reward.

With the wraparound, since we're just using `get_bytes_diff()` and not looking at absolute memory usage, each iteration would have approximately the same slope and wouldn't go negative. We're giving Criterion's `iter_custom()` approximately `get_bytes_diff()*iters`. This is what the graph looks like:

![image](https://github.com/aws/s2n-tls/assets/76919968/a9811506-0c44-4a38-b4ec-3d295407702f)

Reply:
Talked offline, decided to remove memory benching from the Criterion report since Criterion isn't designed for memory benching and this way of including it in the report was too hack-y. Memory benching is now separated out into it's own folder with custom graphing, similar to in historical performance: https://github.com/aws/s2n-tls/pull/4083.